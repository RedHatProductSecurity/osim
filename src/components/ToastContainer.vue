<script setup lang="ts">
import Toast from '../components/widgets/Toast.vue';
import { useToastStore } from '../stores/ToastStore';
import { computed } from 'vue';
import { navbarBottom, footerTop, footerHeight } from '@/stores/responsive';
import { useSettingsStore } from '@/stores/SettingsStore';

// const props = defineProps<{
//   // timestamp: moment.Moment,
// }>();

const { toasts } = useToastStore();
const { settings } = useSettingsStore();

const top = computed<string>(() => {
  return navbarBottom.value > 0 ? 'auto' : '0';
});
const position = computed<string>(() => {
  return navbarBottom.value > 0 ? 'absolute' : 'fixed';
});
const height = computed<string>(() => {
  if (navbarBottom.value > 0) {
    return footerTop.value - navbarBottom.value + 'px';
  } else {
    return 'calc(100vh - ' + footerHeight.value + 'px)';
  }
});

function clearAll() {
  console.log('clearing all');
  toasts.length = 0;
  // while (toasts.length > 0) {
  //   toasts.shift();
  // }
}

</script>

<template>
  <!-- <div
    class="osim-toast-container
    toast-container position-fixed top-0
    bottom-0 end-0 overflow-auto overflow-x-hidden p-3"
  > -->
  <div class="osim-toast-container toast-container p-3">
    <div
      class="toast"
      style="display:block;visibility: hidden;height:0;margin:0;padding:0;border:0;"
    ><!--
      Hack: when removing the last toast, the container width is set to 0,
            which causes the toast to clip or squash when being animated.
            This preserves the container width, which fixes animation.
    --></div>
    <Transition name="toast">
      <div
        v-if="settings.showNotifications && toasts.length > 0"
        class="osim-toast-container-clear d-flex justify-content-end"
      >
        <button
          v-if="settings.showNotifications && toasts.length > 0"
          type="button"
          class="btn btn-secondary btn-sm mb-3"
          @click.prevent="clearAll"
        >Clear All<i class="bi bi-trash3 ms-1"></i></button>
      </div>
    </Transition>
    <TransitionGroup name="toast">
      <Toast
        v-for="(toast, index) in toasts"
        v-show="toast.isFresh || settings.showNotifications"
        :key="toast.id"
        :title="toast.title"
        :body="toast.body"
        :bodyHtml="toast.bodyHtml"
        :timestamp="toast.timestamp"
        :timeoutMs="toast.timeoutMs"
        :css="toast.css"
        @close="toasts.splice(index, 1)"
        @stale="toast.isFresh = false"
      >
      </Toast>
    </TransitionGroup>
  </div>
</template>

<style scoped>
.osim-toast-container-clear button {
  pointer-events: auto;
}

.osim-toast-container {
  position: v-bind(position);
  top: v-bind(top);

  /* height: calc(100vh - 100px); */
  height: v-bind(height);
  overflow: clip auto;

  /* clip-path: inset(-100vw 0 -100vw -100vw); */
  right: 0;

  /* justify-self: end; */
}

.osim-toast-container-clear {
  pointer-events: auto;
}

.toast-move,
.toast-enter-active,
.toast-leave-active {
  transition: all 0.2s ease;
}

.toast-enter-from,
.toast-leave-to {
  opacity: 0;
  transform: translateX(30px);
}

.toast-leave-active {
  position: absolute;
}
</style>
