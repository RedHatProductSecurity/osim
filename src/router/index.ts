import {createRouter, createWebHistory} from 'vue-router'
import IndexView from '../views/IndexView.vue'
import LoginView from '../views/LoginView.vue';
import IssueDetails from '../views/IssueDetails.vue';
import FlawView from '../views/FlawView.vue';
import Tracker from '../views/Tracker.vue';
import {useUserStore} from '@/stores/UserStore';

const router = createRouter({
  history: createWebHistory(import.meta.env.BASE_URL),
  routes: [
    {
      path: '/',
      name: 'index',
      component: IndexView,
    },
    {
      path: '/issue-details',
      name: 'issue-details',
      component: IssueDetails,
    },
    {
      path: '/login',
      name: 'login',
      component: LoginView,
      meta: {
        hideNavbar: true,
      },
    },
    {
      path: '/tracker',
      name: 'tracker',
      component: Tracker,
    },

    {
      path: '/flaw-details',
      name: 'flaw-details',
      component: FlawView,
    },
    // {
    //   path: '/about/:id',
    //   name: 'about',
    //   // route level code-splitting
    //   // this generates a separate chunk (About.[hash].js) for this route
    //   // which is lazy-loaded when the route is visited.
    //   component: () => import('../views/AboutView.vue')
    // },
  ]
})


router.beforeEach((to, from) => {
  const {isAuthenticated} = useUserStore();

  if (isAuthenticated()) {
    if (to.name === 'login') {
      return false;
    }
  } else {
    if (to.name !== 'login') {
      return {name: 'login'};
    }
  }
})

export default router
