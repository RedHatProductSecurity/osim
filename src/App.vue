<script setup lang="ts">
import { DateTime } from 'luxon';
import { RouterView } from 'vue-router';
import Navbar from './components/Navbar.vue';
import {
  setup,
  osidbHealth,
  osimRuntime,
  osimRuntimeStatus,
  OsimRuntimeStatus,
} from '@/stores/osimRuntime';

import { onBeforeUnmount, onMounted, ref, watch, watchEffect } from 'vue';
import ToastContainer from '@/components/ToastContainer.vue';
import { useElementBounding } from '@vueuse/core';
import { footerHeight, footerTop } from '@/stores/responsive';

setup();

watch(osimRuntimeStatus, () => {
  if (osimRuntimeStatus.value === OsimRuntimeStatus.READY) {
    updateRelativeOsimBuildDate();
  }
});

const elFooter = ref<HTMLElement | null>(null);
const { top: footerTop_, height: footerHeight_ } = useElementBounding(elFooter);
watchEffect(() => {
  footerTop.value = footerTop_.value;
  footerHeight.value = footerHeight_.value;
});
const relativeOsimBuildDate = ref('');
const updateRelativeOsimBuildDate = () => {
  relativeOsimBuildDate.value =
    DateTime.fromISO(osimRuntime.value.osimVersion.timestamp).toRelative() || '';
};
let buildDateIntervalId: number = -1;
onMounted(() => {
  const ms15Minutes = 15 * 60 * 60 * 1000;
  buildDateIntervalId = setInterval(updateRelativeOsimBuildDate, ms15Minutes);
  updateRelativeOsimBuildDate();
});
onBeforeUnmount(() => {
  clearInterval(buildDateIntervalId);
});
</script>

<template>
  <template v-if="osimRuntimeStatus === OsimRuntimeStatus.READY">
    <header>
      <Navbar v-if="!$route.meta.hideNavbar" />
    </header>
    <div class="osim-content-layered">
      <ToastContainer />
      <RouterView class="osim-page-view" />
    </div>
    <footer ref="elFooter" class="fixed-bottom osim-status-bar">
      <!--TODO add active request count-->
      <!--<div>[ Requests: {{ activeRequestCount }} ]</div>-->
      <div>
        [
        OSIM
        | env: {{ osimRuntime.env }}
        | <span :title="osimRuntime.osimVersion.rev">
          tag: {{ osimRuntime.osimVersion.tag }}
          <span v-if="osimRuntime.osimVersion.dirty">(dirty)</span>
        </span>
        | ts : {{ osimRuntime.osimVersion.timestamp
          .substring(0, osimRuntime.osimVersion.timestamp.indexOf('T')) }}
        ({{ relativeOsimBuildDate }})
        ]
      </div>
      <div>
        [
        OSIDB
        | env: {{ osidbHealth.env }}
        | <span :title="osidbHealth.revision.substring(0, 7)">ver: {{ osidbHealth.version }}</span>
        ]
      </div>
      <div v-if="osimRuntime.error" class="osim-status-osidb-err">
        [err: {{ osimRuntime.error }}]
      </div>
    </footer>
  </template>
  <div v-else-if="osimRuntimeStatus === OsimRuntimeStatus.ERROR" class="osim-backend-error">
    {{ osimRuntime.error }}
  </div>
</template>

<style>
:root {
  --osim-status-bar-height: 24px;
}

/*
.osim-view-wrapper {
  position: relative;
  display: grid;
}
.osim-view-wrapper > * {
  grid-column-start: 1;
  grid-row-start: 1;
  align-items: start;
}
*/

.osim-content-layered {
  display: grid;
  margin-bottom: var(--osim-status-bar-height);
}

.osim-content-layered > * {
  grid-column-start: 1;
  grid-row-start: 1;
  align-items: start;
}

.osim-backend-error {
  max-width: 60em;
  border: 1px solid darkred;
  padding: 1em;
  margin: 1em;
  display: flex;
  justify-content: center;
  align-self: center;
}

.osim-page-view {
  margin-bottom: var(--osim-status-bar-height);
}

.osim-status-bar.fixed-bottom {
  /* position: fixed; */
  background: #efefef;
  border-top: 1px solid #ddd;
  height: var(--osim-status-bar-height);
  display: flex;
  justify-content: flex-end;
  position: fixed;
  bottom: 0;
  left: 0;
  right: 0;
  z-index: 1091; /* --bs-toast-zindex + 1 */
}

.osim-status-bar > * {
  border-left: 1px solid #ddd;
  padding: 0 3px;
}
</style>
