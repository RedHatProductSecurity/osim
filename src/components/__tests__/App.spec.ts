import { VueWrapper, mount } from '@vue/test-utils';
import { describe, it, expect } from 'vitest';
import router from '@/router';
import { createTestingPinia } from '@pinia/testing';
import App from '@/App.vue';
import ToastContainer from '../ToastContainer.vue';

describe('App', async () => {
  let subject: VueWrapper<InstanceType<typeof App>>;
  beforeEach(() => {

    vi.mock('@vueuse/core', () => ({
      useSessionStorage: vi.fn(() => ({
        value: {
          refresh: 'mocked_refresh_token',
          env: 'mocked_env',
          whoami: {
            email: 'test@example.com',
            username: 'testuser',
          },
        },
      })),
      useElementBounding: vi.fn(() => ({
        bottom: 1000,
        height: 100,
        top: 100
      }))
    }));

    vi.mock('jwt-decode', () => ({
      default: vi.fn(() => ({
        sub: '1234567890',
        name: 'Test User',
        exp: Math.floor(Date.now() / 1000) + (60 * 60 * 24 * 365)
      })),
    }));

    vi.mock('@/stores/osimRuntime', async () => {
      return {
        setup: vi.fn(() => { }),
        osimRuntimeStatus: 1,
        osidbHealth: {},
        osimRuntime: {},
        OsimRuntimeStatus: {
          INIT: 0,
          READY: 1,
          ERROR: 2,
        },
      };
    });

    router.push('/');
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  it('renders ToastContainer', async () => {
    const pinia = createTestingPinia({
      createSpy: vitest.fn,
      stubActions: false,
    });
    subject = mount(App, {
      global: {
        plugins: [
          pinia,
          router
        ]
      }
    });
    const toastContainer = subject.findAllComponents(ToastContainer)?.[0];
    expect(toastContainer?.exists()).toBeTruthy();
  });
});
