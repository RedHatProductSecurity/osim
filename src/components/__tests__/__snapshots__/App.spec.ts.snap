// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`app > renders the App component when OSIDB is NOT READY 1`] = `"<div class="osim-backend-error">OSIDB is not ready</div>"`;

exports[`app > renders the App component when OSIDB is READY 1`] = `
"<session-warning-banner-stub></session-warning-banner-stub>
<header>
  <navbar-stub></navbar-stub>
</header>
<div class="osim-content-layered">
  <toast-container-stub></toast-container-stub>
  <router-view-stub class="osim-page-view"></router-view-stub>
</div>
<footer class="fixed-bottom osim-status-bar">
  <!--TODO add active request count-->
  <!--<div>[ Requests: {{ activeRequestCount }} ]</div>-->
  <change-log-stub></change-log-stub>
  <div> [ OSIM | env: test | <span title="1"> tag: 1.0.0</span> | ts : 2024-08-29 () ] </div>
  <div> [ OSIDB | env: test | <span title="1">ver: 1.0.0</span> ] </div>
  <div class="osim-status-osidb-err"> [err: OSIDB is not ready] </div>
</footer>"
`;
